### PR TITLE
Set session options when initializing a basic session

### LINKED ISSUE #41932: Session options are nil. Unable to add sessions and automatic connection switching to an api_only app.
I recently updated sidekiq and I need to use sessions in my api_only rails app as per [6.2.1](https://github.com/mperham/sidekiq/blob/master/Changes.md#621). I've tried to add them back in per the [docs](https://edgeguides.rubyonrails.org/api_app.html#using-session-middlewares), but it doesn't work. If I set `config.api_only = true` to `config.api_only = false` then it all works as expected. I opened an issue in the [sidekiq repo](https://github.com/mperham/sidekiq/issues/4872) and I was directed here.

### Steps to reproduce
`config/application.rb`
```
    config.api_only = true

    # This also configures session_options for use below
    config.session_store :cookie_store, key: '_interslice_session'

    # Required for all session management (regardless of session_store)
    config.middleware.use ActionDispatch::Cookies
    config.middleware.use ActionDispatch::Session::CookieStore, key: '_interslice_session'
    config.middleware.insert_after(ActionDispatch::Cookies, ActionDispatch::Session::CookieStore, key: '_interslice_session')

    config.middleware.use config.session_store, config.session_options

```

### Expected behavior
It should load the session cookie config?

### Actual behavior
It looks like the sessions aren't working and the web part of my app won't boot.

```
F, [2021-04-12T09:20:34.579998 #33] FATAL -- : [389c9875-8447-43d7-9e18-bde804be0f4a]
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] NoMethodError (undefined method `[]=' for nil:NilClass):
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a]
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/request/session.rb:235:in `load!'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/request/session.rb:230:in `load_for_write!'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/request/session.rb:216:in `merge!'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/request/session.rb:19:in `create'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/middleware/session/abstract_store.rb:71:in `prepare_session'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] rack (2.2.3) lib/rack/session/abstract/id.rb:265:in `context'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] rack (2.2.3) lib/rack/session/abstract/id.rb:260:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/middleware/cookies.rb:689:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] rack (2.2.3) lib/rack/static.rb:161:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] rack (2.2.3) lib/rack/builder.rb:244:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] sidekiq (6.2.1) lib/sidekiq/web.rb:108:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] sidekiq (6.2.1) lib/sidekiq/web.rb:113:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/routing/mapper.rb:20:in `block in <class:Constraints>'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/routing/mapper.rb:49:in `serve'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/journey/router.rb:50:in `block in serve'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/journey/router.rb:32:in `each'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/journey/router.rb:32:in `serve'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/routing/route_set.rb:842:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] activerecord (6.1.3.1) lib/active_record/middleware/database_selector.rb:53:in `block in call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] activerecord (6.1.3.1) lib/active_record/middleware/database_selector/resolver.rb:63:in `block (2 levels) in read_from_replica'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] activesupport (6.1.3.1) lib/active_support/notifications/instrumenter.rb:24:in `instrument'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] activerecord (6.1.3.1) lib/active_record/middleware/database_selector/resolver.rb:62:in `block in read_from_replica'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] activerecord (6.1.3.1) lib/active_record/connection_handling.rb:383:in `block (2 levels) in with_role_and_shard'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] activerecord (6.1.3.1) lib/active_record/connection_adapters/abstract/connection_pool.rb:1025:in `while_preventing_writes'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] activerecord (6.1.3.1) lib/active_record/connection_handling.rb:381:in `block in with_role_and_shard'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] activerecord (6.1.3.1) lib/active_record/connection_handling.rb:398:in `swap_connection_handler'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] activerecord (6.1.3.1) lib/active_record/connection_handling.rb:373:in `with_handler'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] activerecord (6.1.3.1) lib/active_record/connection_handling.rb:380:in `with_role_and_shard'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] activerecord (6.1.3.1) lib/active_record/connection_handling.rb:175:in `connected_to'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] activerecord (6.1.3.1) lib/active_record/middleware/database_selector/resolver.rb:61:in `read_from_replica'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] activerecord (6.1.3.1) lib/active_record/middleware/database_selector/resolver.rb:39:in `read'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] activerecord (6.1.3.1) lib/active_record/middleware/database_selector.rb:63:in `select_database'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] activerecord (6.1.3.1) lib/active_record/middleware/database_selector.rb:52:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] rack (2.2.3) lib/rack/etag.rb:27:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] rack (2.2.3) lib/rack/conditional_get.rb:27:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] rack (2.2.3) lib/rack/head.rb:12:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/middleware/callbacks.rb:27:in `block in call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] activesupport (6.1.3.1) lib/active_support/callbacks.rb:98:in `run_callbacks'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/middleware/callbacks.rb:26:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/middleware/executor.rb:14:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/middleware/actionable_exceptions.rb:18:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] appsignal (3.0.2) lib/appsignal/rack/rails_instrumentation.rb:32:in `call_with_appsignal_monitoring'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] appsignal (3.0.2) lib/appsignal/rack/rails_instrumentation.rb:17:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/middleware/debug_exceptions.rb:29:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] railties (6.1.3.1) lib/rails/rack/logger.rb:37:in `call_app'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] railties (6.1.3.1) lib/rails/rack/logger.rb:26:in `block in call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] activesupport (6.1.3.1) lib/active_support/tagged_logging.rb:99:in `block in tagged'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] activesupport (6.1.3.1) lib/active_support/tagged_logging.rb:37:in `tagged'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] activesupport (6.1.3.1) lib/active_support/tagged_logging.rb:99:in `tagged'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] railties (6.1.3.1) lib/rails/rack/logger.rb:26:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/middleware/request_id.rb:26:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] rack (2.2.3) lib/rack/runtime.rb:22:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] activesupport (6.1.3.1) lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/middleware/executor.rb:14:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] rack (2.2.3) lib/rack/sendfile.rb:110:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] actionpack (6.1.3.1) lib/action_dispatch/middleware/host_authorization.rb:92:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] rack-cors (1.1.1) lib/rack/cors.rb:100:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] railties (6.1.3.1) lib/rails/engine.rb:539:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] puma (5.2.2) lib/puma/configuration.rb:248:in `call'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] puma (5.2.2) lib/puma/request.rb:76:in `block in handle_request'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] puma (5.2.2) lib/puma/thread_pool.rb:337:in `with_force_shutdown'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] puma (5.2.2) lib/puma/request.rb:75:in `handle_request'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] puma (5.2.2) lib/puma/server.rb:431:in `process_client'
web-74b87b954b-tqv78 — [389c9875-8447-43d7-9e18-bde804be0f4a] puma (5.2.2) lib/puma/thread_pool.rb:145:in `block in spawn_thread'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] rack (2.2.3) lib/rack/head.rb:12:in `call'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] actionpack (6.1.3.1) lib/action_dispatch/middleware/callbacks.rb:27:in `block in call'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] activesupport (6.1.3.1) lib/active_support/callbacks.rb:98:in `run_callbacks'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] actionpack (6.1.3.1) lib/action_dispatch/middleware/callbacks.rb:26:in `call'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] actionpack (6.1.3.1) lib/action_dispatch/middleware/executor.rb:14:in `call'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] actionpack (6.1.3.1) lib/action_dispatch/middleware/actionable_exceptions.rb:18:in `call'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] appsignal (3.0.2) lib/appsignal/rack/rails_instrumentation.rb:32:in `call_with_appsignal_monitoring'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] appsignal (3.0.2) lib/appsignal/rack/rails_instrumentation.rb:17:in `call'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] actionpack (6.1.3.1) lib/action_dispatch/middleware/debug_exceptions.rb:29:in `call'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] actionpack (6.1.3.1) lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] railties (6.1.3.1) lib/rails/rack/logger.rb:37:in `call_app'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] railties (6.1.3.1) lib/rails/rack/logger.rb:26:in `block in call'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] activesupport (6.1.3.1) lib/active_support/tagged_logging.rb:99:in `block in tagged'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] activesupport (6.1.3.1) lib/active_support/tagged_logging.rb:37:in `tagged'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] activesupport (6.1.3.1) lib/active_support/tagged_logging.rb:99:in `tagged'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] railties (6.1.3.1) lib/rails/rack/logger.rb:26:in `call'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] actionpack (6.1.3.1) lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] actionpack (6.1.3.1) lib/action_dispatch/middleware/request_id.rb:26:in `call'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] rack (2.2.3) lib/rack/runtime.rb:22:in `call'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] activesupport (6.1.3.1) lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] actionpack (6.1.3.1) lib/action_dispatch/middleware/executor.rb:14:in `call'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] rack (2.2.3) lib/rack/sendfile.rb:110:in `call'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] actionpack (6.1.3.1) lib/action_dispatch/middleware/host_authorization.rb:92:in `call'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] rack-cors (1.1.1) lib/rack/cors.rb:100:in `call'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] railties (6.1.3.1) lib/rails/engine.rb:539:in `call'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] puma (5.2.2) lib/puma/configuration.rb:248:in `call'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] puma (5.2.2) lib/puma/request.rb:76:in `block in handle_request'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] puma (5.2.2) lib/puma/thread_pool.rb:337:in `with_force_shutdown'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] puma (5.2.2) lib/puma/request.rb:75:in `handle_request'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] puma (5.2.2) lib/puma/server.rb:431:in `process_client'
web-74b87b954b-j8pbj — [564a856e-9bfb-45e9-ba16-a2fa8d4ed391] puma (5.2.2) lib/puma/thread_pool.rb:145:in `block in spawn_thread'
```

### System configuration
Ruby version: 2.7.2
Rails version: 6.1.3.1
Sidekiq version: 6.2.1
Sidekiq Pro version: 5.2.2

### LINKED ISSUE #42306: Incompatibility between api_only, session middleware and automatic database connection switching
### Steps to reproduce

1. Create a new rails api only application `rails new --api`
2. Configure application for multiple databases as described in https://guides.rubyonrails.org/active_record_multiple_databases.html#setting-up-your-application
3. Add session middleware back into the api only application as described in https://guides.rubyonrails.org/api_app.html#using-session-middlewares
4. Activate automatic connection switching as described in https://guides.rubyonrails.org/active_record_multiple_databases.html#activating-automatic-connection-switching
5. Visiting the welcome page will not fail.

[This repo](https://github.com/jgrau/api-only-with-replica) contains an application with a failing integration test. [The commit history](https://github.com/jgrau/api-only-with-replica/commits/master) might be useful - it follows the same steps as above.

### Expected behavior

Automatic database switching should work

### Actual behavior

This error is raised:

```
❯ BACKTRACE=1 rake
Run options: --seed 34305

# Running:

E

Error:
FooTest#test_can_see_the_welcome_page:
NoMethodError: undefined method `[]=' for nil:NilClass
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/actionpack-6.1.3.2/lib/action_dispatch/request/session.rb:235:in `load!'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/actionpack-6.1.3.2/lib/action_dispatch/request/session.rb:230:in `load_for_write!'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/actionpack-6.1.3.2/lib/action_dispatch/request/session.rb:216:in `merge!'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/actionpack-6.1.3.2/lib/action_dispatch/request/session.rb:19:in `create'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/actionpack-6.1.3.2/lib/action_dispatch/middleware/session/abstract_store.rb:71:in `prepare_session'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/rack-2.2.3/lib/rack/session/abstract/id.rb:265:in `context'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/rack-2.2.3/lib/rack/session/abstract/id.rb:260:in `call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/actionpack-6.1.3.2/lib/action_dispatch/middleware/cookies.rb:689:in `call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/activerecord-6.1.3.2/lib/active_record/middleware/database_selector.rb:53:in `block in call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/activerecord-6.1.3.2/lib/active_record/middleware/database_selector/resolver.rb:63:in `block (2 levels) in read_from_replica'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/activesupport-6.1.3.2/lib/active_support/notifications/instrumenter.rb:24:in `instrument'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/activerecord-6.1.3.2/lib/active_record/middleware/database_selector/resolver.rb:62:in `block in read_from_replica'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/activerecord-6.1.3.2/lib/active_record/connection_handling.rb:388:in `with_role_and_shard'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/activerecord-6.1.3.2/lib/active_record/connection_handling.rb:175:in `connected_to'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/activerecord-6.1.3.2/lib/active_record/middleware/database_selector/resolver.rb:61:in `read_from_replica'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/activerecord-6.1.3.2/lib/active_record/middleware/database_selector/resolver.rb:39:in `read'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/activerecord-6.1.3.2/lib/active_record/middleware/database_selector.rb:63:in `select_database'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/activerecord-6.1.3.2/lib/active_record/middleware/database_selector.rb:52:in `call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/rack-2.2.3/lib/rack/etag.rb:27:in `call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/rack-2.2.3/lib/rack/conditional_get.rb:27:in `call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/rack-2.2.3/lib/rack/head.rb:12:in `call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/actionpack-6.1.3.2/lib/action_dispatch/middleware/callbacks.rb:27:in `block in call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/activesupport-6.1.3.2/lib/active_support/callbacks.rb:98:in `run_callbacks'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/actionpack-6.1.3.2/lib/action_dispatch/middleware/callbacks.rb:26:in `call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/actionpack-6.1.3.2/lib/action_dispatch/middleware/executor.rb:14:in `call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/actionpack-6.1.3.2/lib/action_dispatch/middleware/actionable_exceptions.rb:18:in `call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/actionpack-6.1.3.2/lib/action_dispatch/middleware/debug_exceptions.rb:29:in `call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/actionpack-6.1.3.2/lib/action_dispatch/middleware/show_exceptions.rb:33:in `call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/railties-6.1.3.2/lib/rails/rack/logger.rb:37:in `call_app'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/railties-6.1.3.2/lib/rails/rack/logger.rb:26:in `block in call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/activesupport-6.1.3.2/lib/active_support/tagged_logging.rb:99:in `block in tagged'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/activesupport-6.1.3.2/lib/active_support/tagged_logging.rb:37:in `tagged'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/activesupport-6.1.3.2/lib/active_support/tagged_logging.rb:99:in `tagged'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/railties-6.1.3.2/lib/rails/rack/logger.rb:26:in `call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/actionpack-6.1.3.2/lib/action_dispatch/middleware/remote_ip.rb:81:in `call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/actionpack-6.1.3.2/lib/action_dispatch/middleware/request_id.rb:26:in `call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/rack-2.2.3/lib/rack/runtime.rb:22:in `call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/activesupport-6.1.3.2/lib/active_support/cache/strategy/local_cache_middleware.rb:29:in `call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/actionpack-6.1.3.2/lib/action_dispatch/middleware/executor.rb:14:in `call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/actionpack-6.1.3.2/lib/action_dispatch/middleware/static.rb:24:in `call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/rack-2.2.3/lib/rack/sendfile.rb:110:in `call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/actionpack-6.1.3.2/lib/action_dispatch/middleware/host_authorization.rb:92:in `call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/railties-6.1.3.2/lib/rails/engine.rb:539:in `call'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/rack-test-1.1.0/lib/rack/mock_session.rb:29:in `request'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/rack-test-1.1.0/lib/rack/test.rb:266:in `process_request'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/rack-test-1.1.0/lib/rack/test.rb:119:in `request'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/actionpack-6.1.3.2/lib/action_dispatch/testing/integration.rb:279:in `process'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/actionpack-6.1.3.2/lib/action_dispatch/testing/integration.rb:16:in `get'
    /Users/jgrau/.rbenv/versions/2.6.3/lib/ruby/gems/2.6.0/gems/actionpack-6.1.3.2/lib/action_dispatch/testing/integration.rb:372:in `get'
    /Users/jgrau/src/api-only-with-replica/test/integration/foo_test.rb:5:in `block in <class:FooTest>'
```
### System configuration
**Rails version**: 6.1.3.2

**Ruby version**: ruby 2.6.3p62 (2019-04-16 revision 67580) [x86_64-darwin20]

### PR DESCRIPTION
**Pull request by gmcgibbon**
### Summary

Closes https://github.com/rails/rails/issues/41932

Sets session options when setting a default basic session. Overrides `Rack::Request#session`, which is public.

### CODE REVIEW

File: actionpack/lib/action_dispatch/http/request.rb [hunk 1]
```diff
@@ -365,7 +365,13 @@ def reset_session
       if session && session.respond_to?(:destroy)
         session.destroy
       else
-        self.session = {}
+        default_session
+      end
+    end
+
+    def session
+      fetch_header(Rack::RACK_SESSION) do |k|
```
Comment (byroot) [inline actionpack/lib/action_dispatch/http/request.rb:373]:
```suggestion
      fetch_header(Rack::RACK_SESSION) do
```

**Comment by zzak** [comment]
@gmcgibbon Could you rebase this please :bow:

cc #42306

**Comment by ricardotk002** [comment]
Hey I've been looking into this issue too, and came up with a solution that seems simpler: https://github.com/ricardotk002/rails/commit/70ce3879ec8444c6479bc87b7069b699a6eab67c

Happy to open a PR if you agree.

**Comment by zzak** [comment]
@ricardotk002 I think we should wait to see what Gannon says, but I do like the test you added :)

**Comment by gmcgibbon** [comment]
Personally, I think `|| {}`ing session options might be disadvantageous for two reasons:

- It is too defensive. The code will cover cases where the session options don't exist in _all_ cases.
- It isn't an instance of `ActionDispatch::Request::Session::Options` which may cause future bugs.

That said, I think the test is a good addition. I can add it to this PR and give you credit. Would that be acceptable @ricardotk002?

**Comment by ricardotk002** [comment]
@gmcgibbon I just checked you recent changes and they make sense to me. It'd be nice if you add the test for me, thanks!

**Review by zzak — APPROVED** [review]

**Pull request merged**